### ISSUE TITLE
cmake: RegularExpression::compile(): Error in compile.

### ISSUE THREAD

**Issue by hakonhagland** [trilinos/Trilinos#4206]
I am trying to build trilinos on Ubuntu 18.10 (CMake version 3.12.1, GCC version 8.2.0)
```
git clone https://github.com/trilinos/Trilinos.git
cd Trilinos
mkdir build
cd build
sudo apt-get install libnetcdf-dev libmatio-dev doxygen
pyenv local 2.7.15
cmake \
  -DTPL_ENABLE_MPI=ON \
  -DMPI_BASE_DIR=/usr/lib/x86_64-linux-gnu/openmpi \
  -DTrilinos_ENABLE_ALL_PACKAGES=ON \
  -DCMAKE_INSTALL_PREFIX=/opt/trilinos \
  -D Trilinos_ENABLE_PyTrilinos:BOOL=ON \
  -D BUILD_SHARED_LIBS:BOOL=ON \
  .. # <path to Trilinos source>
```

This fails with 
```
[...]

RegularExpression::compile(): Nested *?+.
RegularExpression::compile(): Error in compile.
CMake Error at cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake:2760 (STRING):
  STRING sub-command REGEX, mode MATCH failed to compile regex
  "/home/hakon/Trilinos/packages/rol".
Call Stack (most recent call first):
  cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake:2803 (TRIBITS_EXCLUDE_FILES)
  packages/rol/CMakeLists.txt:114 (TRIBITS_EXCLUDE_AUTOTOOLS_FILES)

[...]
RegularExpression::compile(): Nested *?+.
RegularExpression::compile(): Error in compile.
CMake Error at packages/PyTrilinos/cmake/UseSWIG.cmake:196 (IF):
  if given arguments:

    "/home/hakon/Trilinos/build/packages/PyTrilinos/src" "MATCHES" "^/home/hakon/Trilinos/packages/PyTrilinos/src"

  Regular expression
  "^/home/hakon/Trilinos/packages/PyTrilinos/src"
  cannot compile
Call Stack (most recent call first):
  packages/PyTrilinos/cmake/UseSWIG.cmake:302 (SWIG_ADD_SOURCE_TO_MODULE)
  packages/PyTrilinos/src/CMakeLists.txt:330 (SWIG_ADD_MODULE)

-- Configuring incomplete, errors occurred!
```

**Comment by wfspotz** [comment]
@hakonhagland I cannot see any reason why

    "^/home/hakon/Trilinos/packages/PyTrilinos/src"

would not compile as a regular expression.

@bartlettroscoe do you see anything here that might be an issue?

**Comment by hakonhagland** [comment]
@wfspotz According to `packages/PyTrilinos/cmake/UseSWIG.cmake:196` the offending test is:
```
  IF(${swig_source_file_path} MATCHES "^${CMAKE_CURRENT_SOURCE_DIR}")
```

shouldn't the first argument to `IF` be quoted? So like this: 
```
  IF("${swig_source_file_path}" MATCHES "^${CMAKE_CURRENT_SOURCE_DIR}")
```

**Comment by wfspotz** [comment]
I believe the syntax is correct as long as `${swig_source_file_path}` is not empty, which it should be.

However, you could test it, if you like.  Be sure to make the same changes to the `IF`, `ELSE` and `ENDIF` statements.

**Comment by hakonhagland** [comment]
I think I found the problem, see

[CMake don’t work when there are certain characters in build directory name](https://gitlab.kitware.com/cmake/cmake/issues/16207)

So I installed `Trilinos` in a folder name with special characters (I am sorry I removed that part of the folder name from the output in my report, but I believed it was not of relevance).

Now I tried again, first installing in `/home/hakon/Trilinos/` and it works fine, whereas
`/home/hakon/Trilinos/c++--trilinos--clone/Trilinos/` did not work. So this might be an indication 
that `++` in the folder name could be the problem.

**Comment by wfspotz** [comment]
I had Googled a similar post as well, but didn't see any regex characters in the file path.

What I saw suggested using `STREQUALS` instead of `MATCHES` and getting rid of the `^` symbol.  If you test it and it works, we can put in a pull request.

**Comment by hakonhagland** [comment]
Ok using `STREQUALS` seems to solve a problem, but there are other errors after changing 
`packages/PyTrilinos/cmake/UseSWIG.cmake:196` from 
```
  IF("${swig_source_file_path}" MATCHES "^${CMAKE_CURRENT_SOURCE_DIR}")
```
to 
```
  IF("${swig_source_file_path}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
```

For example there are a lot of errors coming from another cmake file: `cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake`. These errors are on the form:

```
RegularExpression::compile(): Nested *?+.
RegularExpression::compile(): Error in compile.
CMake Error at cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake:2759 (STRING):
  STRING sub-command REGEX, mode MATCH failed to compile regex
  "/home/hakon/Trilinos/c++/packages/rol".
Call Stack (most recent call first):
  cmake/tribits/core/package_arch/TribitsGlobalMacros.cmake:2802 (TRIBITS_EXCLUDE_FILES)
  packages/rol/CMakeLists.txt:114 (TRIBITS_EXCLUDE_AUTOTOOLS_FILES)
```

In my opinion the real problem is not `MATCHES` function but the `++` in the build folder name. For example, 
above I installed in `/home/hakon/Trilinos/c++`, if I had installed in `/home/hakon/Trilinos` instead, it would have worked fine.. so it had nothing to with `MATCH` and `STREQUAL` as far as I can see. Maybe you could test installing to a folder containing the string `c++`? To check if this error is reproducable.

**Comment by wfspotz** [comment]
I have no doubt that having `++` in the file path is the problem.  The `+` symbol is a regular expression character that means "one or more of the previous characters".  `++` is illegal regular expression syntax.  `MATCHES` is regular expression matching, while `STREQUALS` uses string comparison.  So it all makes sense.

I hadn't seen this before, which means it does not seem to come up that often, but looking at the logic, there is no reason to be using regular expressions when comparing these file paths.  But it also looks like changing from regular expressions to string comparisons won't be trivial, and the entire build system will have to be tested.  I can open an issue about it, but I don't think it will be a high priority.

**Comment by yifangt** [comment]
I happened to have the same issue with my path name with "C++" for one folder. The error message was always "Regular expression" in STRING match etc. Nobody had any issue to play with the [same library](https://github.com/simongog/sdsl-lite) except me. I could not figure out what the problem was until I find this thread. After I changed the "C++" part to "CPP" in the folder name, everything worked just fine! This give another proof that having ++ in the file path can be the problem.
